python/ray/air/execution/resources/fixed.py

```diff
@@ -0,0 +1,149 @@
+from typing import Dict, List, Optional, Type, Union
+
+from dataclasses import dataclass
+
+import ray
+from ray import SCRIPT_MODE, LOCAL_MODE
+from ray.air.execution.resources.request import (
+    ResourceRequest,
+    AllocatedResource,
+)
+from ray.air.execution.resources.resource_manager import ResourceManager
+from ray.util.annotations import DeveloperAPI
+
+
+# Avoid numerical errors by multiplying and subtracting with this number.
+# Compare: 0.99 - 0.33 vs (0.99 * 1000 - 0.33 * 1000) / 1000
+_DIGITS = 100000
+
+
+def _sum_bundle_resources(bundles: List[Dict[str, float]]) -> Dict[str, float]:
+    all_resources = {}
+    for resources in bundles:
+        for k, v in resources.items():
+            all_resources[k] = all_resources.get(k, 0) + v
+
+    return all_resources
+
+
+@DeveloperAPI
+@dataclass
+class FixedAllocatedResource(AllocatedResource):
+    bundles: List[Dict[str, float]]
+
+    def annotate_remote_objects(
+        self, objects: List[Type]
+    ) -> List[Union[ray.ObjectRef, ray.actor.ActorHandle]]:
+        # If we have an empty head, we schedule the first object (the "head") with
+        # empty resources.
+        if self.resource_request.head_bundle_is_empty:
+            bundles = [{}] + self.resource_request.bundles
+        else:
+            bundles = self.resource_request.bundles
+
+        if len(objects) > len(bundles):
+            raise RuntimeError(
+                f"The number of objects to annotate ({len(objects)}) cannot "
+                f"exceed the number of available bundles ({len(bundles)})."
+            )
+
+        annotated = []
+        for i, (obj, bundle) in enumerate(zip(objects, bundles)):
+            bundle = bundle.copy()
+            num_cpus = bundle.pop("CPU", 0)
+            num_gpus = bundle.pop("GPU", 0)
+            memory = bundle.pop("memory", 0.0)
+
+            annotated.append(
+                obj.options(
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class FixedResourceManager(ResourceManager):
```

Comment:
One thing that's a bit confusing to me is how global is the ResourceManager abstraction?

For example, with `PlacementGroupResourceManager`, if I were to create an instance of it, serialize it, and then deserialize it on another node, it will still refer to the same resource manager since it is backed by placement groups underneath the hood. So it is truly global in that sense.

However, this is not the case with FixedResourceManager since the state is all local.

Is this global property something that should be baked into the ResourceManager interface itself and expected of all resource managers? If so, should FixedResourceManager be backed by an actor?

Reply:
Let me know if this is too much bikeshedding

Reply:
I think in the first instance we should just avoid serialization of either. As in, both are stateful, and both will malfunction if they are effectively duplicated.
For sharing a manager, yes, they should live on an actor.

I'll add a serialization prevention.